### DISCUSSION TITLE
Can we get the results from a catch's select statement?

### DISCUSSION THREAD

**Discussion by nishantvarma** [sqlalchemy/sqlalchemy#7982] [Usage Questions] [answered]
I want to select the MSSQL's [xact_state](https://docs.microsoft.com/en-us/sql/t-sql/functions/xact-state-transact-sql?view=sql-server-ver15) variable to know if a transaction is uncommitable:

```
1> BEGIN TRANSACTION;
2> BEGIN TRY
3>   SELECT CAST('A' AS INT);
4> END TRY
5> BEGIN CATCH
6>   PRINT XACT_STATE();
7> END CATCH
8> GO
           
-----------

(0 rows affected)
-1
```

Is there a way to do that? I tried selecting normally, but that doesn't help:

```
qry = """
BEGIN TRY
  SELECT CAST('A' AS INT);
END TRY
BEGIN CATCH
  SELECT XACT_STATE();
END CATCH
"""

# How to detect xact_state = -1 and raise out?
res = session.execute(qry).fetchall();
```

The rationale is to raise out only if the transaction is _uncommittable_; all other cases, I am fine to continue with the next query. (The library is exposing that as a feature, so it is hard to change the behavior.)

**Comment by zzzeek** [comment, marked as answer]
im just guessing from the "GO" youre using SQL Server, but in any case, for imperative SQL programming like stored procs or catches you need to run those operations using the DBAPI connection directly, see https://docs.sqlalchemy.org/en/14/core/connections.html#working-with-the-dbapi-cursor-directly and then if this is pyodbc get further help at https://github.com/mkleehammer/pyodbc

**Reply by nishantvarma** [reply]
Sorry that I didn't specify it's MSSQL. I have clarified it above as well.

**Comment by nishantvarma** [comment]
I asked this in pyODBC as well: https://github.com/mkleehammer/pyodbc/issues/1295. Just for reference.

**Comment by gordthompson** [comment]
On further investigation pursuant to https://github.com/mkleehammer/pyodbc/issues/1295 it appears that pyodbc cannot retrieve the `xact_state()` value correctly, but pymssql can:

```python
from sqlalchemy import Column, Integer, String, create_engine, text
from sqlalchemy.orm import Session, declarative_base

connection_url = "mssql+pymssql://scott:tiger^5HHH@192.168.0.199/test"
engine = create_engine(connection_url)
Base = declarative_base()


class User(Base):
    __tablename__ = "user"
    id = Column(Integer, primary_key=True, autoincrement=False)
    name = Column(String(50))


Base.metadata.drop_all(engine)
Base.metadata.create_all(engine)

with Session(engine) as sess:
    sess.add(User(id=1, name="Gordon"))
    sess.commit()

with Session(engine) as session:
    user_1 = session.get(User, 1)
    user_1.name = "Gord"
    session.flush()
    try:
        session.execute(text("SELECT CAST('A' AS INT)")).fetchall()
    except:
        result = session.scalar(text("select xact_state()"))
        print(result)  # 0 (no transaction in effect)
```

**Comment by nishantvarma** [comment]
So pymmsql is able to detect 0, not -1, unlike plain SQL.

```
 1 means that the transaction is committable.  
-1 means the transaction is uncommittable and should be rolled back.  
 0 means there is no transaction and a commit or rollback operation would generate an error. 
```

**Reply by gordthompson** [reply]
That may be a matter of timing. While the T-SQL batch is still running (in "plain SQL"), the transaction may still exist but be uncommittable (`xact_state()` returns -1). However, when pymssql does `select xact_state()` it would be in a different batch on the same connection. Terminating the previous batch (the one that caused the error) might have already rolled back and discarded the transaction, so `xact_state()` returns 0. I don't know that for sure, but it seems like a reasonable explanation.

And, in fact, if we try calling `session.commit()` after `xact_state()` returned 0 then pymssql throws an exception telling us that "no transaction exists".

**Comment by mkleehammer** [comment]
This is curious.  Since there are no parameters involved, pyodbc will use SQLExecDirect, so it isn't the usual issue of MSSQL using a stored procedure under the covers.  In this example, I would expect them to retrieve the same values.

Can you provide an ODBC trace for each so we could compare?

**Reply by nishantvarma** [reply]
I have replied in the pyODBC thread. I create a a simple test case and capture odbc.log for both scenarios.
